### PR TITLE
Feature consumer authorization

### PR DESCRIPTION
**Pull request by stefanpearson**
This is to give more granular control over what permissions apps (consumers) have access to. For instance, a dashboard app may require access to private API data, but cannot perform any user interactions (as a user may not be associated with the app).

Consumers will now have to be registered in the database with the desired scope permissions (see schema).

To make an API request, sign a JWT with your consumer `secret` and include the consumer `key` and `accessToken` (optional, only required if a user is involved).

```
// Sign this with HS256 and your consumer secret (don't include any expiration date)
{
  "key": "123",
  "accessToken": "abc"
}
```

### CODE REVIEW

**Comment by ribotaccounts** [comment]
Can one of the admins verify this patch?

**Comment by matto1990** [comment]
add to whitelist

File: .env.example [hunk 1]
```diff
@@ -1,5 +1,5 @@
 PORT=4568
-BASE_URL=http://localhost:4568
+BASE_URL=http://localhost:4569
```
Comment (matto1990) [inline .env.example]: We shouldn't need to change this, but it's fine to leave in

File: app/controllers/auth.js [hunk 2]
```diff
@@ -32,7 +35,7 @@ utils.promisify( googlePlus.people );
  */
 var init = function init() {
 
-  passport.use( new passportHttpBearer.Strategy( {}, findPersonByAccessToken ) );
+  passport.use( new passportHttpBearer.Strategy( {}, authenticate ) );
 
   router.post( '/auth/sign-in',
     middleware.validateBody,
```
Comment (matto1990) [inline app/controllers/auth.js]: Should this not require the `user` scope?

File: app/lib/utils.js [hunk 3]
```diff
@@ -103,8 +103,8 @@ var promisify = function promisify( obj ) {
  * Encodes a given payload (objects/arrays will be JSON stringified)
  */
 var encodeToken = function encodeToken( payload ) {
-  var token = jwt.encode( payload, environment.jwtSecret );
-  return token;
+  var signedToken = jwt.sign( payload, environment.jwtSecret );
+  return signedToken;
```
Comment (matto1990) [inline app/lib/utils.js]: Why not just return directly?

File: app/lib/utils.js [hunk 4]
```diff
@@ -112,8 +112,10 @@ var encodeToken = function encodeToken( payload ) {
  * Decodes a token back into a payload
  */
 var decodeToken = function decodeToken( token ) {
-  var decoded = jwt.decode( token, environment.jwtSecret );
-  return decoded;
+  var decodedToken = jwt.verify( token, environment.jwtSecret, {
+    ignoreExpiration: true
+  } );
+  return decodedToken;
```
Comment (matto1990) [inline app/lib/utils.js]: Why not just return directly?

File: app/controllers/auth.js (hunk 2 above)
Comment (stefanpearson) [inline app/controllers/auth.js]: I guess, but the scopes are done through a bearer token, and this doesn't require one. Not sure…

File: data/schema.js [hunk 5]
```diff
@@ -10,6 +10,18 @@ var schema = {
     }
   },
 
+  consumer: {
```
Comment (matto1990) [inline data/schema.js]: You need to increment the schema version at the top of this file for the automatic migrations to happen correctly.

File: test/mocha/helpers/index.js [hunk 6]
```diff
@@ -185,8 +186,20 @@ var helpers = {
...
    */
-  isValidDate: function( date ) {
+  isValidDate: function isValidDate( date ) {
     return moment( date, [ 'ddd, DD MMM YYYY HH:mm:ss ZZ' ] ).isValid();
+  },
+
+  /**
+   * Sign a JWT with a given accessToken record and consumer record
+   */
+  signJwt: function testFn( accessToken, consumer ) {
```
Comment (matto1990) [inline test/mocha/helpers/index.js]: Incorrect function name

**Comment by matto1990** [comment]
I left a few style comments, but there's a slightly larger issue. I don't think this solves the issue that Manuel was having with the API. Namely, he wants to be able to access the check-in data without needing a user access token. I don't think this PR allows that.

To access check-in information you now need to provide:

* A consumer key with the `checkIn` scope
* An access token for a user

We need to remove that last requirement to make it work for the dashboard.

One solution I can think of is the have a consumer with a special scope, something like `checkInUnauthenticated`. In the `checkAuthenticationIfRequired` method in `app/controllers/ribots.js` (https://github.com/ribot/ribot-api/pull/8/files?diff=split#diff-f67bae7bf96c367015b4e75b789140c7R48) we could then check that either the consumer has the `checkIn` scope and an `accessToken` or the consumer has the `checkInUnauthenticated` scope.

That would solve the issue Manuel is having.

Make sense?

**Pull request closed without merging**